### PR TITLE
[flydsl] PA DECODE

### PR DESCRIPTION
**Pull request by ahmed-bsod**
## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

**Review by vgokhale — DISMISSED** [review]

File: aiter/ops/flydsl/kernels/pa_decode_gfx1250.py [hunk 1]
```diff
@@ -0,0 +1,1170 @@
...
+            ):
+                buffer_ops.buffer_store(lo, out_rsrc, off_lo)
+                buffer_ops.buffer_store(hi, out_rsrc, off_hi)
+
+        off_lse = lse_base + lane16
+        if arith.cmpi(
+            arith.CmpIPredicate.slt, _raw(lane16), _raw(qgs_idx)
+        ):
+            buffer_ops.buffer_store(m_state, ml_rsrc, off_lse)
+            buffer_ops.buffer_store(l_state, es_rsrc, off_lse)
```
[the remark this answers is not in the document]
Comment (ahmed-bsod) [reply aiter/ops/flydsl/kernels/pa_decode_gfx1250.py]: verified in the asm and mlir. it gets lowered properly and masking is there

File: aiter/ops/flydsl/pa_decode.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        output.view(-1),
+        tmp_out.view(-1),
+        max_logits.view(-1),
+        exp_sums.view(-1),
+        seq_lens.view(-1),
+        num_seqs,
+        num_kv_heads,
+        num_partitions,
+        stream,
+    )
```
[the remark this answers is not in the document]
Comment (ahmed-bsod) [reply aiter/ops/flydsl/pa_decode.py]: fixed

File: aiter/ops/flydsl/kernels/pa_decode_gfx1250.py [hunk 3]
```diff
@@ -0,0 +1,1197 @@
...
+    # currently supported.
+    if not (1 <= QUERY_GROUP_SIZE <= WMMA_M):
+        raise ValueError(
+            f"QUERY_GROUP_SIZE must be in [1, {WMMA_M}], got {QUERY_GROUP_SIZE}"
+        )
+    if dtype not in ("bf16", "f16"):
+        raise ValueError(f"dtype must be 'bf16' or 'f16', got {dtype!r}")
+
+    BLOCKS_PER_COMPUTE = KV_COMPUTE_BLOCK_SIZE // KV_BLOCK_SIZE
+    # Physical block IDs are loaded via s_buffer_load with vec_width ∈ {1, 2, 4}.
```
[the remark this answers is not in the document]
Comment (ahmed-bsod) [reply aiter/ops/flydsl/kernels/pa_decode_gfx1250.py]: fixed

File: aiter/ops/flydsl/pa_decode.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    ), f"Q head_size {head_size} != KV head_size {head_size_kv}"
+    assert (
+        num_q_heads % num_kv_heads == 0
+    ), f"num_q_heads {num_q_heads} not divisible by num_kv_heads {num_kv_heads}"
+    query_group_size = num_q_heads // num_kv_heads
+
+    assert block_tables.shape[0] == num_seqs
+    assert seq_lens.shape[0] == num_seqs
+    assert block_tables.dtype == torch.int32
+    assert seq_lens.dtype == torch.int32
```
[the remark this answers is not in the document]
Comment (ahmed-bsod) [reply aiter/ops/flydsl/pa_decode.py]: fixed

File: op_tests/flydsl_tests/test_flydsl_pa_decode.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    (128, 32, 16, 1, 1, 384, 96, 192),  # BPC=3, multi-iter (2 computes/partition)
+    (128, 32, 16, 1, 1, 448, 224, 224),  # BPC=7, single-iter, 2 partitions
+]
+
+
+@pytest.mark.parametrize(
+    "head_size,kv_block_size,query_group_size,num_kv_heads,num_seqs,seq_len,kv_compute_block_size,partition_size",
+    CASES,
+)
+@pytest.mark.parametrize("dtype", [torch.bfloat16])
```
[the remark this answers is not in the document]
Comment (ahmed-bsod) [reply op_tests/flydsl_tests/test_flydsl_pa_decode.py:170]: nah

File: op_tests/flydsl_tests/test_flydsl_pa_decode.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+Reference: a simple torch SDPA-style implementation over the paged KV cache.
+"""
+
+import math
+import random
+
+import pytest
+import torch
+
```
[the remark this answers is not in the document]
Comment (ahmed-bsod) [reply op_tests/flydsl_tests/test_flydsl_pa_decode.py:14]: fixed

File: aiter/ops/flydsl/pa_decode.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+
+    device = query.device
+    dtype_str = _dtype_to_str(query.dtype)
+
+    # Allocate tmp buffers for split-KV outputs.
+    tmp_out = torch.empty(
+        (num_seqs, num_kv_heads, num_partitions, query_group_size, head_size),
+        dtype=torch.float32,
+        device=device,
+    )
```
[the remark this answers is not in the document]
Comment (ahmed-bsod) [reply aiter/ops/flydsl/pa_decode.py:225]: fixed

File: aiter/ops/flydsl/pa_decode.py [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+    if not value_cache.is_contiguous():
+        raise ValueError("value_cache must be contiguous")
+    if not output.is_contiguous():
+        raise ValueError("output must be contiguous")
+    if not block_tables.is_contiguous():
+        raise ValueError("block_tables must be contiguous")
+    if not seq_lens.is_contiguous():
+        raise ValueError("seq_lens must be contiguous")
+
+    max_seq_len = int(seq_lens.max().item()) if num_seqs > 0 else 0
```
[the remark this answers is not in the document]
Comment (ahmed-bsod) [reply aiter/ops/flydsl/pa_decode.py]: fixed

File: aiter/ops/flydsl/kernels/pa_decode_gfx1250.py [hunk 9]
```diff
@@ -0,0 +1,1211 @@
...
+                valid = valid_next
+
+            # `m_state` is not needed in the epilogue.
+            l_final = l_state
+            acc_final = acc_vec
+
+            inv_l = rocdl.rcp(T.f32, l_final)
+            inv_l_vec = vector.broadcast(vec_f32_ty, inv_l)
+            out_vec_f32 = arith.mulf(acc_final, inv_l_vec)
+            out_vec_half = arith.trunc_f(vec_half_ty, out_vec_f32)
```
[the remark this answers is not in the document]
Comment (ahmed-bsod) [reply aiter/ops/flydsl/kernels/pa_decode_gfx1250.py]: fixed

File: aiter/ops/flydsl/kernels/pa_decode_gfx1250.py [hunk 10]
```diff
@@ -0,0 +1,1224 @@
...
+stage. 2-stage double buffer overlaps the next tile's loads with current compute.
+"""
+
+import flydsl.compiler as flyc
+import flydsl.expr as fx
+
+from flydsl._mlir import ir
+from flydsl._mlir.dialects import llvm as _llvm
+from flydsl.compiler.kernel_function import CompilationContext
+from flydsl.expr import arith, buffer_ops, gpu, range_constexpr, rocdl, tdm_ops, vector
```
Comment (coderfeli) [inline aiter/ops/flydsl/kernels/pa_decode_gfx1250.py]: This kernel introduces a lot of direct dialect-level usage in the main kernel body (`arith.index`, `arith.index_cast`, `arith.addf`/`mulf`/`maximumf`/`select`, `vector.extract`/`from_elements`/`broadcast`/`shuffle`, plus runtime control flow that will lower to `scf`). Please keep the low-level ops at true boundaries such as `rocdl`, and TDM/resource descriptor code, but use the FlyDSL internal types/helpers for ordinary indexing, arithmetic, softmax updates, and vector packing (`fx.Index`, `fx.Int32`, `ArithValue` operators, `cond.select(...)`, `Vec` helpers, and `fx.range(..., init=...)`).

**Comment by coderfeli** [comment]
Was it tested on silicon and worked ok?

File: aiter/ops/flydsl/kernels/pa_decode_gfx1250.py [hunk 11]
```diff
@@ -0,0 +1,1224 @@
...
+    w1 = arith.trunci(i32_t, base_hi_i64)
+
+    # word 2: num_records (bytes)
+    w2 = _raw(arith.constant(num_records_bytes, type=T.i32))
+
+    # word 3: flags (data format, OOB select, etc.)
+    w3 = _raw(arith.constant(buffer_ops._get_buffer_flags(arch), type=T.i32))
+
+    rsrc_type = ir.Type.parse("vector<4xi32>")
+    return vector.from_elements(rsrc_type, [w0, w1, w2, w3])
```
Comment (fsx950223) [inline aiter/ops/flydsl/kernels/pa_decode_gfx1250.py]: Use FlyDSL ops?

File: aiter/ops/flydsl/pa_decode.py [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+        arch = ""
+    arch_base = arch.lower().split(":")[0] if arch else ""
+    if not arch_base.startswith("gfx1250"):
+        raise ValueError(
+            f"flydsl_paged_attention_decode requires gfx1250, got {arch!r}"
+        )
+
+    # --- Shape derivations + GQA / head-size checks ---
+    num_seqs, num_q_heads, head_size = query.shape
+    num_blocks, num_kv_heads, kv_block_size, head_size_kv = key_cache.shape
```
[the remark this answers is not in the document]
Comment (ahmed-bsod) [reply aiter/ops/flydsl/pa_decode.py]: i already added a max_seq_len==0 check. it just returns output of zero in that case without launching the kernel

**Comment by ahmed-bsod** [comment]
> Was it tested on silicon and worked ok?

@coderfeli yes and also tested with flydsl==0.1.8

File: aiter/ops/flydsl/kernels/pa_decode_gfx1250.py (hunk 11 above)
Comment (ahmed-bsod) [reply aiter/ops/flydsl/kernels/pa_decode_gfx1250.py]: @fsx950223 I couldn't find a flydsl op for scalar loads `s_buffer_load` similar to `buffer_ops.buffer_load` so had to add all this.

File: aiter/ops/flydsl/kernels/pa_decode_gfx1250.py (hunk 10 above)
Comment (ahmed-bsod) [reply aiter/ops/flydsl/kernels/pa_decode_gfx1250.py]: @coderfeli I made the requested changes to the code. lmk if there's more that needs to be changed

File: aiter/ops/flydsl/kernels/pa_decode_gfx1250.py [hunk 13]
```diff
@@ -0,0 +1,1179 @@
...
+        )
+        q_lds.get()
+        k_lds.get()
+        v_lds.get()
+
+        # --- qk_scale in log2 domain (so we can use exp2) ---
+        qk_scale_f32 = fx.Float32(arith.bitcast(T.f32, i32_qk_scale.ir_value()))
+        LOG2E = fx.Float32(1.4426950408889634)
+        qk_scale_log2_scalar = qk_scale_f32 * LOG2E
+        qk_scale_log2_vec = vector.broadcast(
```
Comment (coderfeli) [inline aiter/ops/flydsl/kernels/pa_decode_gfx1250.py]: too many raw mlir vector  usage. Could you use internal Vector type to replace it?

File: aiter/ops/flydsl/kernels/pa_decode_gfx1250.py [hunk 14]
```diff
@@ -0,0 +1,1179 @@
...
+                    phys_i32 = arith.select(in_range, phys_i32, zero_i32)
+                    out.append(fx.Index(phys_i32))
+                else:
+                    # Vectorized scalar K$ load. b64 (width 2) and b128 (width 4)
+                    # are the hardware-native sizes.
+                    phys_vec = _s_buffer_load_vec(
+                        bt_rsrc_v4i32, bt_off_bytes_i32, this_width
+                    )
+                    for b in range_constexpr(this_width):
+                        elem = vector.extract(
```
Comment (coderfeli) [inline aiter/ops/flydsl/kernels/pa_decode_gfx1250.py]: Vector

File: aiter/ops/flydsl/kernels/pa_decode_gfx1250.py [hunk 15]
```diff
@@ -0,0 +1,1179 @@
...
+            # Scale by qk_scale * log2e so we can use exp2 later.
+            for n_tile in range_constexpr(N_QK_TILES):
+                qk_accs[n_tile] = qk_accs[n_tile] * qk_scale_log2_vec
+
+            # Col mask (token < seq_len)
+            neg_inf_f32_local = fx.Float32(float("-inf"))
+            for n_tile in range_constexpr(N_QK_TILES):
+                new_vals = []
+                for mi in range_constexpr(8):
+                    v = vector.extract(
```
Comment (coderfeli) [inline aiter/ops/flydsl/kernels/pa_decode_gfx1250.py]: Vector

File: aiter/ops/flydsl/kernels/pa_decode_gfx1250.py [hunk 16]
```diff
@@ -0,0 +1,1179 @@
...
+                        neg_finite_max_vec8,
+                    )
+
+            # ------------------------ Online softmax (1-warp) ------------------------
+            # Per-lane = 8 KVB-tokens × 1 QGSP-row. Reduction is along KVB.
+            # m_state, l_state, alpha are SCALAR per lane (lane16 = row).
+
+            # Step 1: per-lane local max across all (n_tile, mi) → scalar.
+            local_max = fx.Float32(
+                vector.extract(qk_accs[0], static_position=[0], dynamic_position=[])
```
Comment (coderfeli) [inline aiter/ops/flydsl/kernels/pa_decode_gfx1250.py]: Vector

File: aiter/ops/flydsl/kernels/pa_decode_gfx1250.py [hunk 17]
```diff
@@ -0,0 +1,1179 @@
...
+                # +2*K_OPS prefetched = 4*K_OPS total. Drain 1*K_OPS (= cur K).
+                tdm_ops.tensor_wait(3 * K_OPS_PER_WAVE)
+            else:
+                # In flight: 2*K_OPS (cur K+V, prefetched in prev iter or prologue).
+                # No new prefetch. Drain 1*K_OPS (= cur K), leave V in flight.
+                tdm_ops.tensor_wait(K_OPS_PER_WAVE)
+
+            # ------------------------ QK WMMA (SWAPPED operands) ------------------------
+            qk_accs = [
+                arith.constant_vector(0.0, T.vec(8, T.f32)) for _ in range(N_QK_TILES)
```
Comment (coderfeli) [inline aiter/ops/flydsl/kernels/pa_decode_gfx1250.py]: fx.Float32?

File: aiter/ops/flydsl/kernels/pa_decode_gfx1250.py [hunk 18]
```diff
@@ -0,0 +1,1179 @@
...
+        kv_head = gpu.block_id("y")
+
+        warp_id = tx / fx.Index(WAVE_SIZE)
+        lane_id = tx % fx.Index(WAVE_SIZE)
+
+        num_kv_heads = fx.Index(i32_num_kv_heads)
+        num_parts = fx.Index(i32_num_parts)
+        num_q_heads = num_kv_heads * fx.Index(QUERY_GROUP_SIZE)
+
+        sl_rsrc = buffer_ops.create_buffer_resource(arg_seq_lens, max_size=True)
```
Comment (coderfeli) [inline aiter/ops/flydsl/kernels/pa_decode_gfx1250.py:964]: no need oob check?

File: aiter/ops/flydsl/kernels/pa_decode_gfx1250.py [hunk 19]
```diff
@@ -0,0 +1,1179 @@
...
+
+        stride_lse_seq = num_kv_heads * num_parts * fx.Index(QUERY_GROUP_SIZE)
+        stride_lse_head = num_parts * fx.Index(QUERY_GROUP_SIZE)
+        stride_lse_part = fx.Index(QUERY_GROUP_SIZE)
+
+        stride_out_seq = num_q_heads * fx.Index(HEAD_SIZE)
+        stride_out_head = fx.Index(HEAD_SIZE)
+
+        tmp_rsrc = buffer_ops.create_buffer_resource(arg_tmp_out, max_size=True)
+        out_rsrc = buffer_ops.create_buffer_resource(arg_out, max_size=True)
```
Comment (coderfeli) [inline aiter/ops/flydsl/kernels/pa_decode_gfx1250.py:985]: no oob?

File: aiter/ops/flydsl/kernels/pa_decode_gfx1250.py (hunk 18 above)
Comment (ahmed-bsod) [reply aiter/ops/flydsl/kernels/pa_decode_gfx1250.py:964]: oob is impossible here. grid is (num_seqs, num_kv_heads, num_parts) and seq_idx is block_id(x) so it will always be within [0, num_seqs]

File: aiter/ops/flydsl/kernels/pa_decode_gfx1250.py (hunk 19 above)
Comment (ahmed-bsod) [reply aiter/ops/flydsl/kernels/pa_decode_gfx1250.py:985]: these will always be accessed in bounds by design. for tmp_rsrc the index is always < NUM_PARTITOINS and for out_rsrc offset is `out_off = seq_idx*stride_out_seq + (kv_head*QGS + r)*stride_out_head + lane_col_base` and 
  - seq_idx is always [0, num_seqs]
  - r < QGS 
  - lane_col_base + COLS_PER_THREAD is within [0, HEAD]

File: aiter/ops/flydsl/kernels/pa_decode_gfx1250.py (hunk 13 above)
Comment (ahmed-bsod) [reply aiter/ops/flydsl/kernels/pa_decode_gfx1250.py]: @coderfeli  replaced raw mlir usage with internal vector type. please take a look thanks

File: aiter/ops/flydsl/pa_decode.py [hunk 20]
```diff
@@ -0,0 +1,297 @@
...
+        raise ValueError(f"`stream` must be on {device}, got {stream.device}")
+
+    # Pack float32 attn_scale as i32 for kernel arg (kernel bitcasts back).
+    scale_i32 = struct.unpack("<i", struct.pack("<f", float(attn_scale)))[0]
+
+    # Pin the current device so JIT compile + kernel launches go to query.device,
+    # regardless of the caller's current CUDA context.
+    with torch.cuda.device(device):
+        # Compile kernels (cached)
+        main_launch = compile_pa_decode_main(
```
Comment (coderfeli) [inline aiter/ops/flydsl/pa_decode.py:251]: this may run jit closure everytime and large cpu overhead. Add flyc. compile and a cache?

Comment (ahmed-bsod) [reply aiter/ops/flydsl/pa_decode.py:251]: addressed in the latest commit
